proposals/0000-componentized-cabal.rst

```diff
@@ -0,0 +1,336 @@
+   package.  This flag is only valid when a component is explicitly
+   specified to be built. For backwards compatibility, the ``--ipid`` flag
+   retains its original meaning as an identifier for the package as a
+   whole: thus when we build a component with it, the IPID is qualified
+   with the name of the component to form a component id.
+
+
+4. If a library contains a ``build-tools`` dependency on an executable, if
+   the library is being configured by itself, it is expected that the
+   caller arrange for the executable to be present on the ``PATH``, so
+   that the build tool configuration process succeeds. (This currently
+   happens automatically for internal executables.)
+
+5. While a user can still explicitly specify installation paths for
+   various files the component may install, we will apply different
+   defaults to these paths::
+
+        Variable    Current         New
+        -----------------------------------------
+        bindir      $prefix/bin     (same)
+        libsubdir   $abi/$libname   $abi/$cid
+        datasubdir  $abi/$pkgid     $abi/$cid
+        docdir      $datadir/doc/$abi/$pkgid     $datadir/doc/$abi/$cid
+
+6. As you might expect, subsequent ``build``, ``register``, etc commands
+   operate only on the enabled component.
+
+There are a few expectations as to how this per-component configure
+interface is intended to be used by a client:
+
+1. If the client reuses the same source directory to build multiple
+   components, it must assign a distinct build directory ``--distdir`` to
+   each component build.
+
+2. The installation directories configured for each component are
+   expected to be disjoint for each component. This is encouraged by the
+   different default install directories. 
+
+3. A component must be copied and registered to a package database
+   within the package database, before any of its (same-package)
+   dependents can be built.
+
+4. The ordering components are built should respect internal
+   ``build-tools`` dependencies.
+
+**Interaction with hooks.** This proposal leaves the hooks API
+unchanged. Thus, just as before, hooks are responsible for determining
+what components are enabled before performing the operations they need.
+If they operate on the available ``ComponentLocalBuildInfo`` hooks should
+automatically work properly with these changes (even better than
+manually specifying which components to build at ``./Setup build`` time)
+
+**Interaction with Haddock.** Initially, ``haddock`` will be kept as it is
+today: a command which builds Haddock documentation for every enabled
+component (one in this case). However, eventually, it would be good to
+treat building Haddock documentation as a "component" in and of itself.
+A Haddock build target for a library has a single dependency on the
+actually built library, making it easy to retroactively build the
+Haddock documentation for a package after the fact.
```

Comment:
OK, there's a new paragraph about this. I haven't looked at the Haddock bug, though I doubt the proposed patch fixes it.